Merged all open PR's request into one pull request

jkwakman opened the pull request:
Bundled all changes into one PR. Merging and maintaining all the separate PR's causes a mess.

jkwakman committed:
Upgraded to PHPUnit 6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
PHPUnit v6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
PHPUnit v6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
PHPUnit v6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
PHPUnit v6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
PHPUnit v6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
PHPUnit v6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
PHPUnit v6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
PHPUnit v6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
PHPUnit v6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
PHPUnit v6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
PHPUnit v6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
PHPUnit v6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
PHPUnit v6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
PHPUnit v6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
PHPUnit v6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
PHPUnit v6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
PHPUnit v6 namespace

PHPUnit v5 namespace is deprecated. Changed it to the new v6 namespace.

jkwakman committed:
Bumped phpunit to version 6

Current version of phpunit. Support untill 1 feb 2019.

jkwakman committed:
Added a test for a field alias

Here we test if the field Alias is working as expected.

jkwakman committed:
Added fieldAlias option

When using the OnetoMany relation we often do not want the parameter name as the fieldname in the Solr database. For example..

Company -> Orders ($orders) -> getOrderNumber()

Here the fieldname will be casted to orders_ss in the solr_database. Instead we want order_id_ss.

To arrange this behaviour we added the fieldAlias in the annotation.

 * @Solr\Field(type="strings", getter="getOrderId", fieldAlias="order_id")

jkwakman committed:
Added docs for the fieldAlias tag

To define an alias field name for the database entry you can use the fieldAlias in the @Solr\Field tag.

jkwakman committed:
Cleanup of type

jkwakman committed:
Added a solr db example

Added a db example after code review

jkwakman committed:
Renamed costom to custom

Added typo fix from costom to custom.

jkwakman committed:
Renamed costom to custom

Added typo fix from costom to custom.

jkwakman committed:
Renamed costom to custom

Added typo fix from costom to custom.

jkwakman committed:
Renamed costom to custom

Added typo fix from costom to custom.

jkwakman committed:
Renamed costom to custom

Added typo fix from costom to custom.

jkwakman committed:
Renamed costom to custom

Added typo fix from costom to custom.

jackabovo committed:
This fix adds a compatibility layer for ext-mongo (SolrBundle uses Doctrine MongoDB ODM) so you can run the application on PHP-7 on which ext-mongo will not run.

jackabovo committed:
Added PHPUnit pretty result printer. This is a modern, pretty printer and gives better feedback compared to the original PHPUnit result printer.

jackabovo committed:
Downgraded phpunit-result-printer to 0.8.0 for PHP 7 compat

jackabovo committed:
Added PHPUnit pretty result printer PHP7.0 version.

jackabovo committed:
Codedungeon has some messed up dependencies for PHP7.0. Trying the right version for the result printer.

jkwakman committed:
Added test entity for defining multiple fields

jkwakman committed:
Rename ValidTestEntityWithMultpileFields to ValidTestEntityWithMultipleFields

jkwakman committed:
Added tests for multiple fields

jkwakman committed:
Test for fields tag getter & remove parenthesis

jkwakman committed:
Added Fields element

jkwakman committed:
Added fields support

Field now had a fields getter and fieldAlias property to support the multi field feature.

jkwakman committed:
Added fields file for the fields annotation support

jkwakman committed:
Added fields annotation to the document factory

jkwakman committed:
Added multiple fields option to the docs

jkwakman committed:
Added @Solr\Fields annotation example

jkwakman committed:
Added namespace for valid test entity

jkwakman committed:
Added valid test entity for multiple fields

jkwakman committed:
Delete ValidTestEntityWithMultipleFields

jkwakman committed:
Merge pull request #1 from jkwakman/added_mongo_php_adapter

This fix adds a compatibility layer for ext-mongo (SolrBundle uses Do…

jkwakman committed:
Merge pull request #2 from jkwakman/jkwakman-field-alias-doctrine-relation

Jkwakman field alias doctrine relation

jkwakman committed:
Merge branch 'pr-collection-branche' into jkwakman-one-to-many-relations

jkwakman committed:
Merge pull request #3 from jkwakman/jkwakman-one-to-many-relations

Jkwakman one to many relations

jkwakman committed:
Merge pull request #4 from jkwakman/jkwakman-phpunit-6

Jkwakman phpunit 6

jkwakman committed:
Merge pull request #5 from jkwakman/jkwakman-typo-refactor-costom

Jkwakman typo refactor costom

jkwakman committed:
Added } in Test

floriansemm commented:
I will test your pull-request in the end of the week

floriansemm commented on Doctrine/Annotation/AnnotationReader.php:
```diff
@@ -31,36 +33,105 @@ public function __construct(Reader $reader)
-            $property->setAccessible(true);
-            $annotation->value = $property->getValue($entity);
-            $annotation->name = $property->getName();
+            if($type == $this::FIELDS_CLASS) {
```
missing space between `if` and bracket

floriansemm commented on Doctrine/Annotation/AnnotationReader.php:
```diff
@@ -31,36 +33,105 @@ public function __construct(Reader $reader)
+                
+                $fields = $this->processFieldsAnnotation($property, $annotation, $entity, $fields);
+            }
+            else {
```
remove new line

jkwakman replied:
Dear Florian,

Thank you for you're time reviewing my PR. I have added the following changes (see commits):

Added fix for inserting array of objects instead of doctrine annotation data.
Added test testReadPropertiesMultipleFields in AnnotationReaderTest.php
Added extra tests for exceptions:
testGetFields_NoParentGetter
testGetFields_UnkownParentGetterMethod
testGetFields_NoFieldAlias
testGetFields_NoFieldGetter
testGetFields_UnknownFieldGetter

Removed newlines and added missing spaces.
Can you please reevaluate my PR? If you require any further information, feel free to contact me.

Kind regards,

Jack Kwakman

floriansemm reviewed (changes requested):
There is a bug in the annotation reader class. I have add locally a test case

```php
    /**
     * @test
     */
    public function readPropertiesFromFields()
    {
        $entity = new EntityWithFields();

        $nested = new ValidTestEntity();
        $nested->setId(rand(1, 10000));
        $nested->setTitle('title 123');

        $entity->setFields([$entity]);

        $this->reader->getFields($entity);
    }
    // ....
class EntityWithFields
{
    /**
     * @Solr\Fields(getter="getFields", fields={
     *      @Solr\Field(type="integers", getter="getId", fieldAlias="id"),
     *      @Solr\Field(type="strings", getter="getTitle", fieldAlias="title")
     *      })
     */
    private $fields;

    /**
     * @return mixed
     */
    public function getFields()
    {
        return $this->fields;
    }

    /**
     * @param mixed $fields
     */
    public function setFields($fields)
    {
        $this->fields = $fields;
    }
}
```

This causes an exception `Unknown method "getId" configured` which I would not expect here.

floriansemm commented on Doctrine/Annotation/AnnotationReader.php:
```diff
@@ -31,36 +33,105 @@ public function __construct(Reader $reader)
+                           $field->getter = $method;
+
+                        } else {
+                            throw new AnnotationReaderException(sprintf('Unknown method "%s" configured', $method));
```
add class in the exception message

jackabovo committed:
Added fix for inserting array of objects instead of doctrine annotation layer.  PR #188

jackabovo committed:
Added exception tests for the annotation reader

jackabovo committed:
Renamed relation to fields (same as example test from Florian)

jkwakman commented:
Dear Florian,

Thank you for you're time reviewing my PR. I have added the following changes (see commits):

- Added fix for inserting array of objects instead of doctrine annotation data.
- Added test testReadPropertiesMultipleFields in AnnotationReaderTest.php
- Added extra tests for exceptions:

testGetFields_NoParentGetter
testGetFields_UnkownParentGetterMethod
testGetFields_NoFieldAlias
testGetFields_NoFieldGetter
testGetFields_UnknownFieldGetter

- Removed newlines and added missing spaces.

Can you please reevaluate my PR? If you require any further information, feel free to contact me. 

Kind regards,

Jack Kwakman

floriansemm commented:
I have tested your feature with an `Post` entity that has a `tags` property.
```
    /**
     * @Solr\Fields(getter="getTags", fields={
     *      @Solr\Field(type="integers", getter="getId", fieldAlias="id"),
     *      @Solr\Field(type="strings", getter="getName", fieldAlias="name")
     *      })
     *
     * @ORM\OneToMany(targetEntity="Acme\DemoBundle\Entity\Tag", mappedBy="post", cascade={"persist", "remove"})
     */
    private $tags;
```

When I persist a post then an error is logged:

`solr.DEBUG: Unknown method defined "getId" in class "Acme\DemoBundle\Entity\Post"`
`Post` has an `getId` method, but I would expect that `Tag::getId` is called in this case.

```php
        $post = new Post();
        $post->setTitle('post with fields');
        $post->setText('text');

        $tag1 = new Tag();
        $tag1->setName('tag #1');

        $tag2 = new Tag();
        $tag2->setName('tag #2');

        $post->setTags([$tag1, $tag2]);
        // save $post comes here....
```
Did I have misunderstood the documentation?

jackabovo committed:
Reevaluated docs example and applied patches for getter call's in parent object.

jackabovo committed:
Removed some extra newlines around the if statements

jkwakman commented:
Dear Florian,

I have reevaluated the changes with the example in the docs. I have e-mailed the example entities so we can be sure we test the same setup. These entities are working without errors on my side and update the Solr database with the following result:

{
        "id":"post_15",
        "solr_type_t":"post",
        "id_is":[29,30],
        "name_ss":["tag #1","tag #2"],
        "_version_":1615765069600129024}]
  }
The tag id's are grouped in the id_is property and the names are grouped under the name_ss. This setup also works for @ManyToOne relationships and can be defined with the same annotation syntax. 

Can you please (re)reevaluate my PR? If you require any further information, feel free to contact me.

Kind regards,

Jack Kwakman

jackabovo committed:
Dear Florian,

Thank you for checking my changes. The problem lies in the fact that the current logic for the EntityIndexerSubsriber binds to the postPersist event. I have changes this to the onFlush event so we have access to all the relational data. Now the id's get persisted as expected.

Changes per file:

AnnotationReader: Allow empty relations to be nullable

DocumentFactory: FieldValue should not be required

EntityIndexerSubscriber: Changed the indexer to onFlush instead of postPersist. At that point the ORM relations are available.

Can you please (re)reevaluate my PR? If you require any further information, feel free to contact me.

Kind regards,

Jack Kwakman

jkwakman committed:
Merge branch 'master' into dev-master

jackabovo committed:
Dear Florian,

Thank you for checking my changes. The problem lies in the fact that the current logic for the EntityIndexerSubsriber binds to the postPersist event. I have changes this to the onFlush event so we have access to all the relational data. Now the id's get persisted as expected.

Changes per file:

AnnotationReader: Allow empty relations to be nullable

DocumentFactory: FieldValue should not be required

EntityIndexerSubscriber: Changed the indexer to onFlush instead of postPersist. At that point the ORM relations are available.

Can you please (re)reevaluate my PR? If you require any further information, feel free to contact me.

Kind regards,

Jack Kwakman

jackabovo committed:
Dear Florian,

Thank you for checking my changes. The problem lies in the fact that the current logic for the EntityIndexerSubsriber binds to the postPersist event. I have changes this to the onFlush event so we have access to all the relational data. Now the id's get persisted as expected.

Changes per file:

AnnotationReader: Allow empty relations to be nullable

DocumentFactory: FieldValue should not be required

EntityIndexerSubscriber: Changed the indexer to onFlush instead of postPersist. At that point the ORM relations are available.

Can you please (re)reevaluate my PR? If you require any further information, feel free to contact me.

Kind regards,

Jack Kwakman

jackabovo committed:
Fixed merge conflicts after remerge with master

jkwakman committed:
Updated treebuilder for SF 4.2

jkwakman committed:
Update FSSolrExtension.php
